internal/gitstream/assign.go

```diff
@@ -0,0 +1,118 @@
+package gitstream
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-git/go-git/v5"
+	"github.com/go-logr/logr"
+	"github.com/google/go-github/v47/github"
+	"github.com/qbarrand/gitstream/internal"
+	"github.com/qbarrand/gitstream/internal/config"
+	gh "github.com/qbarrand/gitstream/internal/github"
+	"github.com/qbarrand/gitstream/internal/gitutils"
+	"github.com/qbarrand/gitstream/internal/markup"
+	"gopkg.in/yaml.v3"
+)
+
+type Assign struct {
+	GC             *github.Client
+	DryRun         bool
+	Finder         markup.Finder
+	GitHelper      gitutils.Helper
+	Logger         logr.Logger
+	IssueHelper    gh.IssueHelper
+	UserHelper     gh.UserHelper
+	Repo           *git.Repository
+	RepoName       *gh.RepoName
+	UpstreamConfig config.Upstream
+}
+
+func (a *Assign) Run(ctx context.Context) error {
+	const remoteName = internal.UpstreamRemoteName
+
+	if _, err := a.GitHelper.RecreateRemote(ctx, remoteName, a.UpstreamConfig.URL); err != nil {
+		return fmt.Errorf("could not recreate remote: %v", err)
+	}
+
+	if err := a.GitHelper.FetchRemoteContext(ctx, remoteName, a.UpstreamConfig.Ref); err != nil {
+		return fmt.Errorf("could not fetch remote %s: %v", remoteName, err)
+	}
+
+	if err := a.assingIssuesAndPRs(ctx); err != nil {
+		return fmt.Errorf("could not add assignees to PRs/Issues: %v", err)
+	}
+
+	return nil
+}
+
+func (a *Assign) getOwnersContent(ctx context.Context) (*gh.Owners, error) {
+
+	const filePath = "OWNERS"
+
+	content, _, _, err := a.GC.Repositories.GetContents(ctx, a.RepoName.Owner, a.RepoName.Repo, filePath, nil)
+	if err != nil || content == nil { // `content` can be nil if the filePath refers to a directory
+		return nil, fmt.Errorf("could not get %s's content from %s/%s: %v", filePath, a.RepoName.Owner, a.RepoName.Repo, err)
+	}
+	data, err := content.GetContent()
+	if err != nil {
+		return nil, fmt.Errorf("%s's data is invalid: %v", filePath, err)
+	}
+
+	var owners gh.Owners
+	if err := yaml.Unmarshal([]byte(data), &owners); err != nil {
+		return nil, fmt.Errorf("could not unmarshal %s as a yaml: %v", filePath, err)
+	}
+
+	return &owners, nil
```

Comment:
Shall we instead return a set of strings containing all the usernames in the file? Faster lookups.

Reply:
Search in a set but longer to pick a random approver and it also add code complexity, therefore, as agreed, we will keep it as is for now and improve it later on if it becomes a bottleneck.